state/factory/factory.go

```diff
@@ -218,18 +251,102 @@ func (sf *factory) NewWorkingSet() (WorkingSet, error) {
 	sf.mutex.RLock()
 	defer sf.mutex.RUnlock()
 
-	return newWorkingSet(
+	return sf.newWorkingSet(
+		context.Background(),
 		sf.currentChainHeight+1,
-		sf.dao,
-		sf.rootHash(),
-		sf.flusherOptions(context.Background(), sf.currentChainHeight+1)...,
 	)
 }
 
+func (sf *factory) newWorkingSet(ctx context.Context, height uint64) (WorkingSet, error) {
+	flusher, err := db.NewKVStoreFlusher(sf.dao, batch.NewCachedBatch(), sf.flusherOptions(ctx, height)...)
+	if err != nil {
+		return nil, err
+	}
+	tlt, err := newTwoLayerTrie(ArchiveTrieRootKey, flusher.KVStoreWithBuffer(), true)
+	if err != nil {
+		return nil, err
+	}
+	if err := tlt.Start(ctx); err != nil {
+		return nil, err
+	}
+	finalized := false
+	trieRoots := make(map[int][]byte)
+
+	return &workingSet{
+		height:    height,
+		finalized: false,
+		getStateFunc: func(ns string, key []byte, s interface{}) error {
+			return readState(tlt, ns, key, s)
+		},
+		putStateFunc: func(ns string, key []byte, s interface{}) error {
+			ss, err := state.Serialize(s)
+			if err != nil {
+				return errors.Wrapf(err, "failed to convert account %v to bytes", s)
+			}
+			flusher.KVStoreWithBuffer().MustPut(ns, key, ss)
+			nsHash := hash.Hash160b([]byte(ns))
+
+			return tlt.Upsert(nsHash[:], key, ss)
+		},
+		delStateFunc: func(ns string, key []byte) error {
+			flusher.KVStoreWithBuffer().MustDelete(ns, key)
+			nsHash := hash.Hash160b([]byte(ns))
+
+			return tlt.Delete(nsHash[:], key)
+		},
+		digestFunc: func() hash.Hash256 {
+			return hash.Hash256b(flusher.SerializeQueue())
+		},
+		finalizeFunc: func(h uint64) error {
+			if finalized {
```

Comment:
imo define all these callback/closure is not as clear as defining methods on a given struct.